esp-hal/src/dma/mod.rs

```diff
@@ -256,6 +250,34 @@ impl DmaDescriptor {
         next: core::ptr::null_mut(),
     };
 
+    /// Resets the descriptor for a new RX transfer.
+    fn reset_for_rx(&mut self) {
+        // Give ownership to the DMA
+        self.set_owner(Owner::Dma);
+
+        // Clear this to allow hardware to set it when the peripheral returns an EOF
+        // bit.
+        self.set_suc_eof(false);
+
+        // Clear this to allow hardware to set it when it's
+        // done receiving data for this descriptor.
+        self.set_length(0);
+    }
+
+    /// Resets the descriptor for a new TX transfer.
+    fn reset_for_tx(&mut self, set_eof: bool) {
+        // Give ownership to the DMA
+        self.set_owner(Owner::Dma);
+
+        // The `suc_eof` bit doesn't affect the transfer itself, but signals when the
+        // hardware should trigger an interrupt request. In circular mode,
+        // we set the `suc_eof` bit for every buffer we send. We use this for
+        // I2S to track progress of a transfer by checking OUTLINK_DSCR_ADDR.
+        // In non-circular mode, we only set it for the last descriptor to signal the
+        // end of the transfer.
```

Comment:
You can now move the `In circular mode, we set .....` part of this comment to the `DescriptorChain` were it matters.

Side note: Setting the EOF bit for every descriptor is excessive but that can be dealt with in a separate PR.

Reply:
I'm pretty much hoping all this flag manipulation is optimised into a single write, or maybe a read-modify-write.